File: Julia/vgg-feature-extractor.jl
function conv_relu(x, out_channels)
    _, _, _, C = size(x)
    W_weights = randn(C, out_channels) .* 0.1
    x_proj = reshape(x, :, C) * W_weights
    x_proj = reshape(x_proj, size(x, 1), size(x, 2), size(x, 3), out_channels)
    max.(0, x_proj)
end

function maxpool_2x2(x)
    B, H, W, C = size(x)
    reshaped = reshape(x, B, div(H, 2), 2, div(W, 2), 2, C)
    maximum(reshaped, dims=(3, 5))
end

function vgg_features(x, config)
    out = x
    for layer in config
        if layer isa Int
            out = conv_relu(out, layer)
        elseif layer == "M"
            out = maxpool_2x2(out)
        end
    end
    out
end

File: R/vgg-feature-extractor.R
conv_relu <- function(x, out_channels) {
  C <- dim(x)[4]
  W_weights <- array(rnorm(C * out_channels) * 0.1, dim = c(C, out_channels))
  x <- apply(x, c(1, 2, 3), function(slice) slice %*% W_weights)
  pmax(0, x)
}

maxpool_2x2 <- function(x) {
  B <- dim(x)[1]
  H <- dim(x)[2]
  W <- dim(x)[3]
  C <- dim(x)[4]
  reshaped <- array(x, dim = c(B, H %/% 2, 2, W %/% 2, 2, C))
  apply(reshaped, c(1, 2, 4, 6), max)
}

vgg_features <- function(x, config) {
  out <- x
  for (layer in config) {
    if (is.numeric(layer)) {
      out <- conv_relu(out, layer)
    } else if (layer == "M") {
      out <- maxpool_2x2(out)
    }
  }
  out
}
